pilz_control/include/pilz_control/traj_mode_manager.h

```diff
@@ -0,0 +1,249 @@
+
+#include <array>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <condition_variable>
+
+namespace pilz_joint_trajectory_controller
+{
+
+enum class TrajProcessingMode
+{
+  //! The controller accepts new trajectories.
+  unhold,
+  //! The controller rejects all new incoming trajectories and stops the robot.
+  stopping,
+  //! The robot stands still and the controller rejects all new incoming trajectories.
+  hold
+};
+
+// "Simple" linear state machine: (start with stopping in accordance to the controller)
+//  x ->  |->  stopping  ->  hold  ->  unhold ->|
+//        |<-               <-                <-|
+static constexpr unsigned int NUM_MODES {3};
+static constexpr std::array<TrajProcessingMode, NUM_MODES> mode_state_machine_{
+  TrajProcessingMode::stopping, TrajProcessingMode::hold, TrajProcessingMode::unhold
+};
+
+inline static constexpr TrajProcessingMode getMode(const unsigned int& idx)
+{
+  return mode_state_machine_[idx];
+}
+
+inline static constexpr bool isTransitionValid(const TrajProcessingMode& requested_new_mode,
+                                               const unsigned int& idx_of_request_new_mode)
+{
+  return requested_new_mode == getMode(idx_of_request_new_mode);
+}
+
+inline static constexpr unsigned int getNextIndex(const unsigned int& current_idx)
+{
+  return (current_idx+1) % NUM_MODES;
+}
+
+/**
+ * @brief Listener to wait for a specified mode to be reached.
+ */
+class TrajProcessingModeListener
+{
+public:
+   //! @param mode Mode to wait for.
+  TrajProcessingModeListener(const TrajProcessingMode& mode);
+  void waitForMode();
+  //! @returns true if the given mode corresponds to the target mode, otherwise false.
+  bool isTargetModeReached(const TrajProcessingMode& mode) const;
+  void triggerListener();
+
+private:
+  std::mutex mutex_;
+  std::condition_variable cond_variable_;
+  bool cond_fulfilled_ {false};
+  //! @brief Mode to wait for.
+  const TrajProcessingMode mode_;
+};
+
+/**
+ * @brief Encapsulates a state machine managing the current Trajectory-Processing-Mode.
+ */
+class TrajProcessingModeManager
+{
+public:
+  //! @returns true only if a successful state switch to stopping was performed, otherwise false.
+  bool stoppingEvent();
+  //! @brief Switch to hold.
+  void stopMotionFinishedEvent();
+  //! @returns true if in state unhold or a successful switch to state unhold was performed, otherwise false.
+  bool unholdEvent();
+
+public:
+  //! @returns true if in state stopping or hold, otherwise false.
+  bool isHolding();
+  bool isUnhold();
+  void registerListener(TrajProcessingModeListener* const listener);
+
+  TrajProcessingMode getCurrentMode();
+
+private:
+  bool switchTo(const TrajProcessingMode& mode, const bool success_at_transition_only=true);
+  bool setMode(const TrajProcessingMode& mode, const bool& success_at_transition_only);
+  void callListener(const TrajProcessingMode& mode);
+  TrajProcessingMode getCurrentModeLockFree() const;
```

Comment:
When trying to add docu I noticed that I cannot explain the difference between `switchTo` and `setMode` in a meaningful way. I know the difference from looking at the code. It does not make much sense to me. I would suggest to refactor `setMode` into a "standard" setter function in order to improve readability.

Sorry for not carefully reviewing this earlier.

More problems I see:
- potential threading problem between the two function calls:
```
  if (setMode(mode, success_at_transition_only))
  {
    callListener(mode);
```
Do you think we should protect this execution by acquiring a lock?

- The parameter `success_at_transition_only` seems redundant to me and should be omitted.

Reply:
> potential threading problem between the two function calls:

I rather suggest that we separate the call to the listeners to an extra thread which has a input queue. This way no state switch message will be lost and no threading problem should occur, given that the input queue of the thread is thread safe. 
If you introduce an additional lock, you will probably make room for a deadlock situation because both functions operate with different locks.

Reply:
> The parameter success_at_transition_only seems redundant to me and should be omitted.

Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.

Reply:
> When trying to add docu I noticed that I cannot explain the difference between switchTo and setMode in a meaningful way. I know the difference from looking at the code. It does not make much sense to me. I would suggest to refactor setMode into a "standard" setter function in order to improve readability.

The decision to separate the `setMode()` and the `callListener()` method, has a lot to do with threading and deadlocks. You can of course suggest something else, but please be careful with the two locks because if you mix them, you will probably make room for deadlocks.

Reply:
> > The parameter success_at_transition_only seems redundant to me and should be omitted.
> 
> Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.

Ok, but why so complicated? The caller of `unholdEvent()` has the possibility to ask `isUnhold()` beforehand. I think you could reduce the complexity of the code by making these event functions more consistent and get rid of `success_at_transition_only`. Decide for yourself.

Reply:
> > potential threading problem between the two function calls:
> 
> I rather suggest that we separate the call to the listeners to an extra thread which has an input queue. This way no state switch message will be lost and no threading problem should occur, given that the input queue of the thread is thread safe.
> If you introduce an additional lock, you will probably make room for a deadlock situation because both functions operate with a different lock.

Hm. I am not really glad with the fact, that whilst one event is processed, another one can be processed in between. The input queue also does not prevent this. As far as I understand the point of a state machine is to prevent such a thing.
Maybe we should rethink the state machine used here. I suggest to consider:
![IMG_9193](https://user-images.githubusercontent.com/35950815/77749718-c6849c80-7022-11ea-8de5-377de099fcca.JPG)

Reply:
> > > The parameter success_at_transition_only seems redundant to me and should be omitted.
> > 
> > 
> > Why is it redundant? The function `unholdEvent()` needs it, to operate correctly.
> 
> Ok, but why so complicated? The caller of `unholdEvent()` has the possibility to ask `isUnhold()` beforehand. I think you could reduce the complexity of the code by making these event functions more consistent and get rid of `success_at_transition_only`. Decide for yourself.

Because the `unholdEvent()` function has to be an atomic operation which includes the `isUnhold()` function call. The `isUnhold()` call cannot be performed externally because then it the complete operation is not atomic.